File: LeetCode/Java/112_path_sum.java

/*
112. Path Sum

Given a binary tree and a sum, determine if the tree has a root-to-leaf path such that adding up all the values along the path equals the given sum.

*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
        private boolean isequal = false; 

    public boolean hasPathSum(TreeNode root, int sum) {
        calculate(root, 0, sum);
        return isequal;
    }

    
    private void calculate(TreeNode node, int cur, int sum) {
        if (!isequal && node != null) { 


            if (node.left == null && node.right == null && (node.val + cur == sum) ) {
                isequal = true;
            }


            if (node.left != null) {
                calculate(node.left, cur + node.val, sum);
            }

            if (node.right != null) {
                calculate(node.right, cur + node.val, sum);
            }
        }
    }
    
}

File: LeetCode/C++/112_Path_Sum.cpp
//112. Path Sum
/*
Given a binary tree and a sum, determine if the tree has a root-to-leaf path such that adding up all the values along the path equals the given sum.

For example:
Given the below binary tree and sum = 22,
              5
             / \
            4   8
           /   / \
          11  13  4
         /  \      \
        7    2      1
return true, as there exist a root-to-leaf path 5->4->11->2 which sum is 22.

Tag: Tree, Depth-first Search

Author: Xinyu Liu
*/

#include <iostream>
#include <vector>
using namespace std;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if (!root)
            return false;
        if (!(root->left) && !(root->right))
            return sum - root->val == 0;
        return hasPathSum(root->left, sum - root->val) || hasPathSum(root->right, sum - root->val);
    }
};

void main(){
    TreeNode n1(1), n2(2), n3(3);
    n1.left = &n2;
    n1.right = &n3;
    Solution sol;
    bool b = sol.hasPathSum(&n1, 6);
}
